File: laundry-reservation/typescript/tests/ReservationServiceBuilder.ts
import { MachineApi } from '../src/MachineApi.js';
import { ReservationService } from '../src/ReservationService.js';
import { FixedMachineSelector } from './FixedMachineSelector.js';
import { FixedPinGenerator } from './FixedPinGenerator.js';
import { InMemoryReservationRepository } from './InMemoryReservationRepository.js';
import { RecordingEmailNotifier } from './RecordingEmailNotifier.js';
import { RecordingMachineDevice } from './RecordingMachineDevice.js';
import { RecordingSmsNotifier } from './RecordingSmsNotifier.js';

export class ReservationServiceBuilder {
  private machineNumber = 7;
  private pins: number[] = [12345];
  readonly device = new RecordingMachineDevice();
  readonly emailNotifier = new RecordingEmailNotifier();
  readonly smsNotifier = new RecordingSmsNotifier();
  readonly repository = new InMemoryReservationRepository();

  withMachineNumber(n: number): this { this.machineNumber = n; return this; }
  withPins(...pins: number[]): this { this.pins = pins; return this; }
  withDeviceRejectingLock(): this { this.device.shouldAcceptLock = false; return this; }

  build(): { service: ReservationService; machineApi: MachineApi } {
    const machineApi = new MachineApi();
    machineApi.registerDevice(this.machineNumber, this.device);

    const service = new ReservationService(
      this.repository,
      this.emailNotifier,
      this.smsNotifier,
      machineApi,
      new FixedPinGenerator(...this.pins),
      new FixedMachineSelector(this.machineNumber),
    );

    return { service, machineApi };
  }
}

File: laundry-reservation/python/tests/reservation_service_builder.py
from __future__ import annotations

from laundry_reservation import MachineApi, ReservationService

from .fixed_machine_selector import FixedMachineSelector
from .fixed_pin_generator import FixedPinGenerator
from .in_memory_reservation_repository import InMemoryReservationRepository
from .recording_email_notifier import RecordingEmailNotifier
from .recording_machine_device import RecordingMachineDevice
from .recording_sms_notifier import RecordingSmsNotifier


class ReservationServiceBuilder:
    def __init__(self) -> None:
        self._machine_number = 7
        self._pins: list[int] = [12345]
        self.device = RecordingMachineDevice()
        self.email_notifier = RecordingEmailNotifier()
        self.sms_notifier = RecordingSmsNotifier()
        self.repository = InMemoryReservationRepository()

    def with_machine_number(self, n: int) -> ReservationServiceBuilder:
        self._machine_number = n
        return self

    def with_pins(self, *pins: int) -> ReservationServiceBuilder:
        self._pins = list(pins)
        return self

    def with_device_rejecting_lock(self) -> ReservationServiceBuilder:
        self.device.should_accept_lock = False
        return self

    def build(self) -> tuple[ReservationService, MachineApi]:
        machine_api = MachineApi()
        machine_api.register_device(self._machine_number, self.device)

        service = ReservationService(
            repository=self.repository,
            email_notifier=self.email_notifier,
            sms_notifier=self.sms_notifier,
            machine_api=machine_api,
            pin_generator=FixedPinGenerator(*self._pins),
            machine_selector=FixedMachineSelector(self._machine_number),
        )

        return service, machine_api
